frontend/server/blockSerialization.ts

```diff
@@ -0,0 +1,309 @@
+    logger.error(message);
+    throw new ServerError(message, HttpStatus.BAD_REQUEST, null);
+  }
+
+  try {
+    return await fs.readFile(absoluteFilePath, { encoding: "utf8" });
+  } catch (error: any) {
+    const message = `Failed to read file: ${absoluteFilePath}`;
+    logger.error(error, message);
+    throw new ServerError(message, HttpStatus.INTERNAL_SERVER_ERROR, error);
+  }
+}
+
+export async function updateBlockFile(
+  pipelinePath: string,
+  blockId: string,
+  relativeFilePath: string,
+  content: string,
+): Promise<void> {
+  const absoluteFilePath = path.join(pipelinePath, blockId, relativeFilePath);
+  const fileName = path.basename(absoluteFilePath);
+  const { write } = getFilePermissions(fileName);
+
+  if (!write) {
+    const message = `Writing file: ${absoluteFilePath}, is not permitted`;
+    logger.error(message);
+    throw new ServerError(message, HttpStatus.BAD_REQUEST, null);
+  }
+
+  try {
+    await fs.writeFile(absoluteFilePath, content);
+  } catch (error: any) {
+    const message = `Failed to write to file: ${absoluteFilePath}`;
+    logger.error(error, message);
+    throw new ServerError(message, HttpStatus.INTERNAL_SERVER_ERROR, error);
+  }
+}
+
+interface FilePermissions {
+  read: boolean;
+  write: boolean;
+}
+
+export function getFilePermissions(name: string): FilePermissions {
+  const readOnly = READ_ONLY_FILES.includes(name);
+  const supported = isFileSupported(name);
+
+  const read = supported;
+  const write = supported && !readOnly;
+
+  return { read, write };
+}
+
+function isFileSupported(filePath: string): boolean {
+  const extension = path.extname(filePath).toLowerCase();
+  const basename = path.basename(filePath);
+
+  return (
+    SUPPORTED_FILE_EXTENSIONS.includes(extension) ||
+    SUPPORTED_FILE_NAMES.includes(basename)
+  );
+}
+
+export async function callAgent(
+  userMessage: string,
+  agentName: string,
+  conversationHistory: any, // Consider creating a specific type for conversation history
+  apiKey: string,
+): Promise<any> {
+  let agents = "agents";
+  if (app.isPackaged) {
+    agents = path.join(process?.resourcesPath ?? "", "agents");
+  }
+  const scriptPath = path.join(
+    agents,
+    agentName,
+    "generate",
+    "computations.py",
+  );
+
+  try {
+    const stdout = await spawnAsync("python", [scriptPath], {
+      input: JSON.stringify({
+        apiKey,
+        userMessage,
+        conversationHistory,
+      }),
+      encoding: "utf8",
+    });
+
+    const parsedOutput = JSON.parse(stdout);
+
+    // Check if the response contains an error
+    if (parsedOutput.error) {
+      const errorDetails: any = parsedOutput.error; // Consider creating a specific type for error details
+      throw new ServerError(
+        errorDetails.message,
+        errorDetails.status_code || HttpStatus.INTERNAL_SERVER_ERROR,
+        {
+          name: errorDetails.type,
+          message: errorDetails.traceback,
+        },
+      );
```

Comment:
I'd be worried about leaking sensitive info by forwarding the error message coming from the script to the user. 
```suggestion
      throw new ServerError(
        `Unable to call agent ${agentName} with message ${userMessage}`;
        errorDetails.status_code || HttpStatus.INTERNAL_SERVER_ERROR,
        parsedOutput.error
      );
```

Reply:
Alternatively you could do 
```suggestion
      throw new Error(parsedOutput.error);
```
and let it be converted to a ServerError in the catch block, and then you could get rid of 
```
 if (error instanceof ServerError) {
      throw error;
    }
```
in the catch block

Reply:
What type of sensitive information are you worried about leaking? It's the users code, running in a pod, on their infrastructure? The error traces are unhelpful if they don't contain the actual traces

Reply:
You can see the error on the console since it runs the script there, this is just surfacing the error to the interface that's being used